proof/refine/ARM/Ipc_R.thy

```diff
@@ -2146,130 +2146,144 @@ lemma replyPush_corres:
+   ignore: threadSet
+   simp: crunch_simps)
+
+lemma refillReady_sp:
+  "\<lbrace>P\<rbrace>
+   refillReady scp
+   \<lbrace>\<lambda>rv s. P s \<and> (\<exists>ko. ko_at' ko scp s \<and> rv = (rTime (refillHd ko) \<le> ksCurTime s + kernelWCETTicks))\<rbrace>"
+  apply (wpsimp wp: refillReady_wp)
+  by fastforce
+
+lemma refillSufficient_sp:
+  "\<lbrace>P\<rbrace>
+  refillSufficient scp k
+  \<lbrace>\<lambda>rv s. P s \<and> (\<exists>ko. ko_at' ko scp s \<and> rv = (sufficientRefills k (scRefills ko) (scRefillHead ko)))\<rbrace>"
+  unfolding refillSufficient_def getRefills_def
+  apply (wpsimp wp: )
+  by (clarsimp simp: sufficientRefills_def obj_at'_def)
+
+lemma isValidTimeoutHandler_sp:
+  "\<lbrace>P\<rbrace>
+   isValidTimeoutHandler x
+   \<lbrace>\<lambda>rv s. P s \<and> (\<exists>ko. ko_at' ko x s \<and> rv = (is_EndpointCap (cteCap (tcbTimeoutHandler ko))))\<rbrace>"
+  unfolding isValidTimeoutHandler_def
+  apply (wpsimp wp: getTCB_wp)
+  by (auto simp: is_EndpointCap_def)
+
+lemma exs_valid_assert_opt [wp]:
+    "\<lbrace> \<lambda>s. G \<noteq> None \<and> Q (the G) s \<rbrace> assert_opt G \<exists>\<lbrace> Q \<rbrace>"
+  by (clarsimp simp: assert_opt_def exs_valid_def get_def
+           assert_def bind_def' return_def)
+
+lemma exs_valid_gets_the [wp]:
+    "\<lbrace> \<lambda>s. h s \<noteq> None \<and> Q (the (h s)) s \<rbrace> gets_the h \<exists>\<lbrace> Q \<rbrace>"
+  by (wpsimp simp: gets_the_def)
+
+lemma valid_tcb_state'_simps[simp]:
+  "valid_tcb_state' Inactive s"
+  "valid_tcb_state' Restart s"
+  by (clarsimp simp: valid_tcb_state'_def)+
+
+lemma state_relation_tcb_relation:
+  "\<lbrakk>(s, s') \<in> state_relation; ko_at (TCB tcb) x s; ko_at' ko x s'\<rbrakk>
+    \<Longrightarrow> tcb_relation tcb ko"
+  apply (clarsimp simp: state_relation_def pspace_relation_def obj_at_def)
+  apply (drule_tac x=x in bspec, clarsimp)
+  apply (clarsimp simp: other_obj_relation_def obj_at'_real_def ko_wp_at'_def projectKOs)
+  done
+
+lemma state_relation_sc_relation:
+  "\<lbrakk>(s, s') \<in> state_relation;
+    ko_at (Structures_A.SchedContext sc n) x s; ko_at' ko x s';
+    valid_sched_context_size n\<rbrakk>
+   \<Longrightarrow> sc_relation sc n ko"
+  apply (clarsimp simp: state_relation_def pspace_relation_def obj_at_def)
+  apply (drule_tac x=x in bspec, clarsimp)
+  apply (clarsimp simp: other_obj_relation_def obj_at'_real_def ko_wp_at'_def projectKOs)
+  done
+
+abbreviation refills_map_precond where
+  "refills_map_precond start count mx list \<equiv> 0 < count \<and> mx \<le> length list \<and> start < mx"
+
+lemma hd_wrap_slice:
+  "refills_map_precond start count mx list \<Longrightarrow> hd (wrap_slice start count mx list) = list ! start"
+  by (auto simp: wrap_slice_def hd_drop_conv_nth)
+
+lemma hd_refills_map:
+  "refills_map_precond start count mx list
+   \<Longrightarrow> hd (refills_map start count mx list) = refill_map (list ! start)"
```

Comment:
This one, too. A lemma that is basically the same as this exists in StateRelation.thy.